UniversalAlgebra/Complexity/FiniteCSP.lagda

```diff
@@ -0,0 +1,198 @@
+date : 2021-07-26
+author: [agda-algebras development team][]
+---
+
+### Constraint Satisfaction Problems
+
+\begin{code}
+
+{-# OPTIONS --without-K --exact-split --safe #-}
+
+module Complexity.FiniteCSP  where
+
+open import Agda.Primitive        using ( _⊔_ ; lsuc ; Level) renaming ( Set to Type )
+open import Agda.Builtin.List     using (List; []; _∷_)
+open import Data.Bool             using ( Bool ; T? ; true)
+open import Data.Bool.Base        using ( T ; _∧_ )
+open import Data.Fin.Base         using ( Fin ; toℕ )
+open import Data.List.Base        using ( length ; [_] ; _++_; head ; tail ; all) renaming (lookup to get)
+open import Data.Product          using ( _,_ ; Σ-syntax ; _×_ )
+open import Data.Vec              using ( Vec ; lookup ; count ; tabulate )
+open import Data.Vec.Relation.Unary.All using ( All )
+open import Data.Nat              using ( ℕ )
+open import Function.Base         using ( _∘_ )
+open import Relation.Binary       using ( DecSetoid ; Rel )
+open import Relation.Nullary      using ( Dec )
+open import Relation.Unary        using ( Pred ; _∈_ ; Decidable )
+
+private variable
+ α ℓ ρ : Level
+
+\end{code}
+
+
+#### Constraints
+
+We start with a vector (of length n, say) of variables.  For simplicity, we'll use the natural
+numbers for variable symbols, so
+
+V : Vec ℕ n
+V = [0 1 2 ... n-1]
+
+We can use the following function to construct the vector of variables.
+
+\begin{code}
+
+range : (n : ℕ) → Vec ℕ n
+range n = tabulate toℕ
+-- `range n` is 0 ∷ 1 ∷ 2 ∷ … ∷ n-1 ∷ []
+
+\end{code}
+
+Let `nvar` denote the number of variables.
+
+A *constraint* consists of
+
+1. a scope vector,  `s : Vec Bool nvar` , where `s i` is `true` if variable `i` is in the scope
+   of the constraint.
+
+2. a constraint relation, rel, which is a collection of functions mapping
+   variables in the scope to elements of the domain.
+
+\begin{code}
+
+-- syntactic sugar for vector element lookup
+_⟨_⟩ : {A : Set α}{n : ℕ} → Vec A n → Fin n → A
+v ⟨ i ⟩ = (lookup v) i
+
+
+open DecSetoid
+
+-- {nvar : ℕ}{dom : Vec (DecSetoid α ℓ) nvar} → 
+
+open Dec
+
+record Constraint {ρ : Level} (nvar : ℕ) (dom : Vec (DecSetoid α ℓ) nvar) : Type (α ⊔ lsuc ρ) where
+ field
+  s   : Vec Bool nvar        -- entry i is true iff variable i is in scope
+  rel : Pred (∀ i → Carrier (dom ⟨ i ⟩)) ρ -- some functions from `Fin nvar` to `dom`
+  dec : Decidable rel
+
+ -- scope size (i.e., # of vars involved in constraint)
+ ∣s∣ : ℕ
+ ∣s∣ = count T? s
+
+ -- point-wise equality of functions when restricted to the scope
+ _≐s_ : Rel (∀ i → Carrier (dom ⟨ i ⟩)) ℓ
+ f ≐s g = ∀ i → T (s ⟨ i ⟩) → (dom ⟨ i ⟩)._≈_ (f i) (g i)
+
+ satisfies : (∀ i → Carrier (dom ⟨ i ⟩)) → Type (α ⊔ ℓ ⊔ ρ) -- An assignment f of values to variables
+ satisfies f = Σ[ g ∈ (∀ i → Carrier (dom ⟨ i ⟩)) ]         -- *satisfies* the constraint provided
+                ( (g ∈ rel) × f ≐s g )                        -- the function f, evaluated at each variable
```

Comment:
why not just `f ∈ rel` ?

Reply:
I can't remember why not... but there was a reason... I'll look at it again.  (The other PR includes that as well... which is why I closed this one... sorry for the confusion!)